core/type/base_param_struct.ex

```diff
@@ -0,0 +1,403 @@
+        end
+    end
+  end
+
+  defunp fetch_from_dict(
+           dict :: [{atom | String.t(), term}] | %{(atom | String.t()) => term},
+           key :: atom,
+           accepted_keys :: [atom]
+         ) :: {:ok, term} | :error do
+    _dict, _key, [] ->
+      :error
+
+    dict, key, [accepted_key | rest] ->
+      case try_fetch_from_dict(dict, accepted_key) do
+        {:ok, value} -> {:ok, value}
+        :error -> fetch_from_dict(dict, key, rest)
+      end
+  end
+
+  defunp try_fetch_from_dict(
+           dict :: [{atom | String.t(), term}] | %{(atom | String.t()) => term},
+           key :: atom
+         ) :: {:ok, term} | :error do
+    dict, key when is_list(dict) ->
+      key_str = Atom.to_string(key)
+
+      Enum.find_value(dict, :error, fn
+        {k, v} when k == key or k == key_str -> {:ok, v}
+        _ -> nil
+      end)
+
+    dict, key when is_map(dict) ->
+      case Map.fetch(dict, key) do
+        {:ok, _} = result -> result
+        :error -> Map.fetch(dict, Atom.to_string(key))
+      end
+  end
+
+  defunp validate_field(value :: term, mod :: v[module]) :: R.t(term, validate_error_t) do
+    if valid_field?(value, mod), do: {:ok, value}, else: {:error, {:invalid_value, [mod]}}
+  end
+
+  @doc false
+  defun valid_field?(value :: term, mod :: v[module]) :: boolean do
+    if :code.get_mode() == :interactive do
+      true = Code.ensure_loaded?(mod)
+    end
+
+    cond do
+      function_exported?(mod, :valid?, 1) -> mod.valid?(value)
+      function_exported?(mod, :__struct__, 0) -> is_struct(value, mod)
+    end
+  end
+
+  @doc false
+  defun update_impl(
+          s :: struct,
+          struct_mod :: module,
+          fields :: [{atom, [atom], module}],
+          dict :: term
+        ) :: R.t(struct, validate_error_t) do
+    s, struct_mod, fields, dict
+    when is_struct(s, struct_mod) and (is_list(dict) or is_map(dict)) ->
+      Enum.map(fields, fn {field_name, accept_field_names, mod} ->
+        fetch_and_validate_field(dict, field_name, accept_field_names, mod)
+      end)
+      |> Enum.reject(&match?({:error, {:value_missing, _}}, &1))
+      |> R.sequence()
+      |> case do
+        {:ok, kvs} ->
+          {:ok, struct(s, kvs)}
+
+        {:error, {reason, mods}}
+        when reason in [:invalid_value, :value_missing] and is_list(mods) ->
+          {:error, {reason, [struct_mod | mods]}}
+      end
+
+    s, struct_mod, _, _ when is_struct(s, struct_mod) ->
+      {:error, {:invalid_value, [struct_mod]}}
+  end
+
+  defmacro __using__(opts) do
+    quote bind_quoted: [opts: opts] do
+      if opts[:accept_case] not in [nil, :snake, :lower_camel, :upper_camel, :capital] do
+        raise ":accept_case option must be one of :snake, :lower_camel, :upper_camel, or :capital"
+      end
+
+      default_pp =
+        Keyword.get(opts, :default_preprocessor, fn _mod -> {:error, :no_default_preprocessor} end)
```

Comment:
Certainly the option did not reflect its role.
I'd prefer to rename to `:preprocessor_generator` because the word `default` could be ambiguous (default preprocessor vs. default generator).
Fixed at d3b77bb.

Reply:
I think you can change `Preprocessor.default/1` as well, maybe, to `PreprocessorGenerator.generate/1` or something.

Reply:
Renamed to `PreprocessorGenerator.generate/1` at a676746 and 65a1f2a.